fem/estimators.cpp

```diff
@@ -214,108 +214,79 @@ void KellyErrorEstimator::ComputeEstimates()
                continue;
             }
 
-            IntegrationRule eir;
-            Vector jumps(nip);
+            // Computation of the transformations is not super cheap, so we delay it until here. All ops above are cheaper lookups.
+            auto FT = mesh->GetFaceElementTransformations(fi);
+
+            // TODO Refactor this computation into a user-facing function.
+            auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(fi));
+            const auto nip = int_rule.GetNPoints();
+
+            double jump_integral = 0.0;
 
             // Integral over local half face on the side of e₁
             // i.e. the numerical integration of ∫ flux ⋅ n dS₁
             for (int i = 0; i < nip; i++)
             {
-               // Evaluate flux at IP
+               // Setup integration point
                auto &fip = int_rule.IntPoint(i);
-               IntegrationPoint ip;
-               FT->Loc1.Transform(fip, ip);
-
-               Vector val(flux_space->GetVDim());
-               flux->GetVectorValue(FT->Elem1No, ip, val);
-
-               // And build scalar product with normal
-               Vector normal(mesh->SpaceDimension());
                FT->Face->SetIntPoint(&fip);
+
+               // Compute weighted normal - note that the normals match at each face integration point up to the sign!
                if (mesh->Dimension() == mesh->SpaceDimension())
                {
                   CalcOrtho(FT->Face->Jacobian(), normal);
                }
                else
                {
-                  Vector ref_normal(mesh->Dimension());
                   FT->Loc1.Transf.SetIntPoint(&fip);
                   CalcOrtho(FT->Loc1.Transf.Jacobian(), ref_normal);
                   auto &e1 = FT->GetElement1Transformation();
                   e1.AdjugateJacobian().MultTranspose(ref_normal, normal);
+                  // We have to cancel the additional weighting from the
+                  // reference to spatial transformation in the line above.
                   normal /= e1.Weight();
                }
-               jumps(i) = val * normal * fip.weight * FT->Face->Weight();
-            }
 
-            // Subtract integral over half face of e₂
-            // i.e. the numerical integration of ∫ flux ⋅ n dS₂
-            for (int i = 0; i < nip; i++)
-            {
-               // Evaluate flux vector at IP
-               auto &fip = int_rule.IntPoint(i);
-               IntegrationPoint ip;
-               FT->Loc2.Transform(fip, ip);
+               // Evaluate flux jump at IP on element 1
+               FT->Loc1.Transform(fip, ip);
+               flux->GetVectorValue(FT->Elem1No, ip, val);
+               double jump = val * normal;
 
-               Vector val(flux_space->GetVDim());
+               // Evaluate flux jump at IP on element 2
+               FT->Loc2.Transform(fip, ip);
                flux->GetVectorValue(FT->Elem2No, ip, val);
+               jump -= val * normal;
```

Comment:
@TobiasDuswald interestingly, these lines already fail. `flux->GetVectorValue(FT->Elem1No, ip, val);` returns slightly different values than `flux->GetVectorValue(FT->Elem2No, ip, val);` for the corresponding transformed integration point. I think the transformation is slightly inexact resulting in the differences here. This leads to some accumulate for higher order elements because they have just more quadrature points per face to integrate properly.

Reply:
I think I'm missing the context here: When you say higher order elements here, do you mean higher solution order, i.e. P, or higher geometry order i.e. curved elements and Q? How big is "slightly different"? The transformation matrices should be consistent for vectors within the tangent space defined by the face. Definitely for linear elements, (and imo for curved too, but I concede that's much harder to pull off).